### PR TITLE
Migrate to the modern linker input API.

### PR DESCRIPTION
**Pull request by benjaminp**
See https://github.com/bazelbuild/bazel/issues/10860.

### CODE REVIEW

**Review by illicitonion — APPROVED** [review]
Thanks!

Do you know what the minimum version of Bazel this supports is?

**Comment by benjaminp** [comment]
3.0.0

**Comment by dfreese** [comment]
docs/index.md mentions 0.26.0, so that should probably be updated.  I don't have a good way of verifying 0.26.0 still works.

**Comment by UebelAndre** [comment]
Does anyone know what it would take to remove `legacy_cc_starlark_api_shim.bzl `?

**Comment by benjaminp** [comment]
It should probably just be renamed or inlined into the callers as desired.

**Comment by UebelAndre** [comment]
So this pull request addresses the usage of any legacy/deprecated behavior in this file?

**Comment by benjaminp** [comment]
Yes.

**Comment by UebelAndre** [comment]
Would it make sense to move those methods into `rust/private/utils.bzl` then?

**Comment by benjaminp** [comment]
Probably. I can do that in this PR if you like.

**Comment by UebelAndre** [comment]
I think that'd be a nice improvement but I leave it up to @illicitonion and @dfreese 😃

**Comment by UebelAndre** [comment]
@damienmg do you have time during your day today to check this out?
@benjaminp If the "legacy" aspect of the modified files no longer applies, could you move the code to `rust/private/utils.bzl`?

File: docs/index.md [hunk 1]
```diff
@@ -36,7 +36,7 @@ load("@io_bazel_rules_rust//:workspace.bzl", "rust_workspace")
 rust_workspace()
 ```
 
-The rules are under active development, as such the lastest commit on the master branch should be used. `master` currently requires Bazel >= 0.26.0.
+The rules are under active development, as such the lastest commit on the master branch should be used. `master` currently requires Bazel >= 3.0.0.
```
Comment (damienmg) [inline docs/index.md:39]: I think https://github.com/bazelbuild/rules_rust/blob/master/workspace.bzl#L5 should be updated too.

**Review by damienmg — CHANGES_REQUESTED** [review]

File: rust/private/legacy_cc_starlark_api_shim.bzl [hunk 2]
```diff
@@ -11,8 +11,8 @@ def get_libs_for_static_executable(dep):
     Returns:
         depset: A depset[File]
     """
-    libraries_to_link = dep[CcInfo].linking_context.libraries_to_link
-    return depset([_get_preferred_artifact(lib) for lib in libraries_to_link.to_list()])
+    linker_inputs = dep[CcInfo].linking_context.linker_inputs.to_list()
```
Comment (UebelAndre) [inline rust/private/legacy_cc_starlark_api_shim.bzl:14]: Can you delete this file and move the contents into `util.bzl`?

**Review by UebelAndre — CHANGES_REQUESTED** [review]

File: rust/private/legacy_cc_starlark_api_shim.bzl (hunk 2 above)
Comment (UebelAndre) [reply rust/private/legacy_cc_starlark_api_shim.bzl]: 🙏

**Review by UebelAndre — APPROVED** [review]

**Review by illicitonion — APPROVED** [review]
LGTM

We should probably try to set up CI running the claimed minimum supported version at some point, but that's definitely not a matter for this PR :)

Over to @damienmg to merge!

**Review by damienmg — APPROVED** [review]

**Comment by damienmg** [comment]
> LGTM
> 
> We should probably try to set up CI running the claimed minimum supported version at some point, but that's definitely not a matter for this PR :)

Yes unfortunately this is a large task as the bazel CI is unlikely to do so. Also Bazel 4 is supposed to be a LTS and I would expect they would test rules with LTS for the support duration of the LTS. At which point the minimal version should always probably be the oldest LTS still supported. 

> 
> Over to @damienmg to merge!

Done 👍 
Thanks for the contributions!

**Comment by damienmg** [comment]
Sorry for merging, I spoke too fast. We need to wait a bit more to see if the rebasing worked.

**Pull request merged**